### PR TITLE
feat: adding documentation describing how to configure Vault for mTLS

### PR DESCRIPTION
**Pull request by rodrigorfk**
**Preview**: https://deploy-preview-1390--cert-manager-website.netlify.app/docs/configuration/vault/#accessing-a-vault-server-with-mtls-enforced

This is related to https://github.com/cert-manager/cert-manager/pull/6614

### CODE REVIEW

**Review by wallrj — CHANGES_REQUESTED** [review]
Thanks @rodrigorfk

I like these new docs, but as an inexperienced vault user I really struggled to get everything setup. I'd like even more detail if possible.

Here's what I did:

* Create Vault serving certificate and CA

```sh
step certificate create "Example Server Root CA" server_ca.crt server_ca.key \
  --profile root-ca \
  --not-after=87600h \
  --no-password \
   --insecure
   

step certificate create vault.vault vault.crt vault.key \
  --profile leaf \
  --not-after=8760h \
  --ca ./server_ca.crt \
  --ca-key server_ca.key \
  --no-password \
  --insecure
```

* Create Vault client certificate and CA

```sh
step certificate create "Example Client Root CA" client_ca.crt client_ca.key \
  --profile root-ca \
  --not-after=87600h \
  --no-password \
   --insecure 

step certificate create client.vault vault_client.crt vault_client.key \
  --profile leaf \
  --not-after=8760h \
  --ca ./client_ca.crt \
  --ca-key client_ca.key \
  --no-password \
  --insecure
```

* Create Vault namespace
```sh
kubectl create ns vault
```

* Create Secret containing Vault serving certificate, Vault client certificate (for use by the readiness probe) and Vault client CA cert (the certificate that Vault will use to verify client certificates.
* 
```sh
kubectl create secret generic vault-tls \
  --namespace vault \
  --from-file=server.key=vault.key \
  --from-file=server.crt=vault.crt \
  --from-file=client-ca.crt=client_ca.crt \
  --from-file=client.crt=vault_client.crt \
  --from-file=client.key=vault_client.key
```

* Deploy Vault

```yaml
# vault-values.yaml
global:
   tlsDisable: false
injector:
  enabled: false
server:
  dataStorage:
    enabled: false
  standalone:
    enabled: true
    config: |
      listener "tcp" {
        address = "[::]:8200"
        cluster_address = "[::]:8201"
        tls_disable = false
        tls_client_ca_file = "/vault/tls/client-ca.crt"
        tls_cert_file = "/vault/tls/server.crt"
        tls_key_file = "/vault/tls/server.key"
        tls_require_and_verify_client_cert = true
      }
  extraArgs: "-dev-tls -dev-listen-address=[::]:8202"
  extraEnvironmentVars:
    VAULT_TLSCERT: /vault/tls/server.crt
    VAULT_TLSKEY: /vault/tls/server.key
    VAULT_CLIENT_CERT: /vault/tls/client.crt
    VAULT_CLIENT_KEY: /vault/tls/client.key
  volumes:
    - name: vault-tls
      secret:
        defaultMode: 420
        secretName: vault-tls
  volumeMounts:
    - mountPath: /vault/tls
      name: vault-tls
      readOnly: true

```

```sh
helm upgrade vault hashicorp/vault --install --namespace vault --create-namespace --values vault-values.yaml
```

* Configure Vault server for Kubernetes auth
```sh
kubectl -n vault exec pods/vault-0  -- \
        vault auth enable --tls-skip-verify kubernetes

kubectl -n vault exec pods/vault-0  -- \
        vault write --tls-skip-verify \
        auth/kubernetes/role/vault-issuer \
        bound_service_account_names=vault-issuer \
        bound_service_account_namespaces=application-1 \
        audience="vault://application-1/vault-issuer" \
        policies=vault-issuer \
        ttl=1m

kubectl -n vault exec pods/vault-0 -- \
        vault write --tls-skip-verify \
        auth/kubernetes/config \
        kubernetes_host=https://kubernetes.default

```

* Create application namespace

```sh
kubectl create ns application-1
```

* Create Service account

```sh
kubectl create serviceaccount -n application-1 vault-issuer
```

* Create Role and Binding

```yaml
# rbac.yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: vault-issuer
  namespace: application-1
rules:
  - apiGroups: ['']
    resources: ['serviceaccounts/token']
    resourceNames: ['vault-issuer']
    verbs: ['create']
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: vault-issuer
  namespace: application-1
subjects:
  - kind: ServiceAccount
    name: cert-manager
    namespace: cert-manager
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: vault-issuer
```

```sh
kubectl apply -f rbac.yaml
```

* Create Issuer

```sh
export CA_BUNDLE=$(base64 -w 0 server_ca.crt)
```

```yaml
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: vault-issuer
  namespace: application-1
spec:
  vault:
    path: pki_int/sign/application-1
    server: https://vault.vault:8200
    caBundle: ${CA_BUNDLE}
    clientCertSecretRef:
      name: vault-client-tls
      key: vault_client.crt
    clientKeySecretRef:
      name: vault-client-tls
      key: vault_client.key
    auth:
      kubernetes:
        role: vault-issuer
        mountPath: /v1/auth/kubernetes
        serviceAccountRef:
          name: vault-issuer
```

```sh
envsubst < vault-issuer.yaml | kubectl -f -
```

* Check Issuer status

```sh
kubectl describe issuer -n application-1
```

File: content/docs/configuration/vault.md [hunk 1]
```diff
@@ -47,6 +47,68 @@ spec:
       ...
 ```
 
+### Accessing a Vault Server with mTLS enforced
+
+In certain use cases, the Vault Server could be configured to enforce clients to present a
+client certificates, those client certificates are just a transport layer enforcement, 
+it does not provide any authentication and authorization mechanism to the Vault APIs itself.
```
[the remark this answers is not in the document]
Comment (rodrigorfk) [reply content/docs/configuration/vault.md]: I add the link to the Vault Server TCP listener configuration, unfortunately hashicorp documentation does not provide a clear description/example on how to require client certificates.

File: content/docs/configuration/vault.md [hunk 2]
```diff
@@ -47,6 +47,68 @@ spec:
...
+```yaml
+apiVersion: cert-manager.io/v1
+kind: Issuer
+metadata:
+  name: vault-issuer
+  namespace: ...
+spec:
+  vault:
+    path: pki_int/sign/example-dot-com
+    server: https://vault.local
```
[the remark this answers is not in the document]
Comment (rodrigorfk) [reply content/docs/configuration/vault.md]: sorry, it was a copy/paste from an example bellow, so it was indeed not realistic and I updated it accordingly to your example

**Comment by rodrigorfk** [comment]
@wallrj , thanks for providing such detailed steps, I have merged it into the documentation, could you please have another look?

**Review by wallrj — CHANGES_REQUESTED** [review]
Thanks.

I simplified some of the file names and used `caBundleSecretRef` to avoid having to base64 encode the ca and use envsubst.

Tested your branch again and it worked.

**Review by wallrj — APPROVED** [review]
Thanks!

/lgtm

**Pull request merged**